src/prof_recent.c

```diff
@@ -0,0 +1,470 @@
+	recent_alloc->alloc_extent = NULL;
+}
+
+/*
+ * This function should be called right before an allocation is released, so
+ * that the associated recent allocation record can contain the following
+ * information:
+ * (1) The allocation is released;
+ * (2) The time of the deallocation; and
+ * (3) The prof_tctx associated with the deallocation.
+ */
+void
+prof_recent_alloc_reset(tsd_t *tsd, extent_t *extent) {
+	/*
+	 * Check whether the recent allocation record still exists without
+	 * trying to acquire the lock.
+	 */
+	if (extent_prof_recent_alloc_get_no_lock(extent) == NULL) {
+		return;
+	}
+
+	prof_tctx_t *dalloc_tctx = prof_tctx_create(tsd);
+	if (dalloc_tctx != NULL) {
+		malloc_mutex_lock(tsd_tsdn(tsd), dalloc_tctx->tdata->lock);
+		++dalloc_tctx->recent_count;
+		dalloc_tctx->prepared = false;
+		malloc_mutex_unlock(tsd_tsdn(tsd), dalloc_tctx->tdata->lock);
+	}
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	prof_recent_t *recent = extent_prof_recent_alloc_get(tsd, extent);
+	if (recent != NULL) {
+		extent_prof_recent_alloc_reset(tsd, extent, recent);
+		assert(nstime_equals_zero(&recent->dalloc_time));
+		assert(recent->dalloc_tctx == NULL);
+		if (dalloc_tctx != NULL) {
+			nstime_update(&recent->dalloc_time);
+			recent->dalloc_tctx = dalloc_tctx;
+		}
+	} else if (dalloc_tctx != NULL) {
+		/* We lost the rase - the allocation record was just gone. */
+		decrement_recent_count(tsd, dalloc_tctx);
+	}
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+}
+
+static void
+prof_recent_alloc_evict_extent(tsd_t *tsd, prof_recent_t *recent) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	if (recent->alloc_extent != NULL) {
+		extent_prof_recent_alloc_reset(tsd, recent->alloc_extent,
+		    recent);
+	}
+}
+
+static void
+assert_prof_recent_alloc_non_empty(tsd_t *tsd) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	assert(prof_recent_alloc_dummy != NULL);
+	assert(prof_recent_alloc_dummy->next != prof_recent_alloc_dummy);
+	assert(prof_recent_alloc_count > 0);
+}
+
+void
+prof_recent_alloc(tsd_t *tsd, extent_t *extent, size_t usize) {
+	assert(extent != NULL);
+	prof_tctx_t *tctx = extent_prof_tctx_get(extent);
+
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), tctx->tdata->lock);
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	prof_recent_t *reserve = NULL;
+	if (prof_recent_alloc_max_load(tsd) == -1 ||
+	    prof_recent_alloc_count < prof_recent_alloc_max_load(tsd)) {
+		assert(prof_recent_alloc_max_load(tsd) != 0);
+		malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		reserve = (prof_recent_t *)iallocztm(tsd_tsdn(tsd),
+		    sizeof(prof_recent_t), sz_size2index(sizeof(prof_recent_t)),
+		    false, NULL, true, arena_get(tsd_tsdn(tsd), 0, false),
+		    true);
+		malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	}
+
+	if (prof_recent_alloc_max_load(tsd) == 0) {
+		assert(prof_recent_alloc_count == 0);
```

Comment:
nit: add a assert function similar to non_empty.

Reply:
Ended up consolidating the two into one function. Also added more assertions using the new function.